### ISSUE TITLE
mount vdi vhd files

### ISSUE THREAD

**Issue by Keops10** [DiscUtils/DiscUtils#268]
Hi I am using Discutils library.
I can read the contents of my files with .vdi-.vdh-.vdhx extensions.
However, I want to copy these files or add them to the computer as a drive and copy them, is it possible?

**Comment by LTRData** [comment]
Not through the DiscUtils library itself. It is possible though for example Arsenal Image Mounter where we use DiscUtils library as backend when you mount certain image file formats.

**Comment by Keops10** [comment]
Yes I tried with Arsenal Imager. However, I think it does not support the file belonging to applications such as virtualbox.

**Comment by LTRData** [comment]
> Yes I tried with Arsenal Imager. However, I think it does not support the file belonging to applications such as virtualbox.

It should support all image formats supported by DiscUtils. Did you see any issues with some format?

**Comment by Keops10** [comment]
Yeah. virtual box does not open image files (.vdi) and bluestacks (.vhdx)

**Comment by LTRData** [comment]
> Yeah. virtual box does not open image files (.vdi) and bluestacks (.vhdx)

I am not sure I follow. I regularly use AIM to mount vdi, vhdx, vmdk etc images myself. It uses DiscUtils library to parse such image files and then another driver to expose them as virtual disks to the system.

**Comment by Keops10** [comment]
you misunderstood me. I can read with DiscUtils but not mount with Arsenal

**Comment by LTRData** [comment]
> you misunderstood me. I can read with DiscUtils but not mount with Arsenal

Yes and that is what surprised me. Arsenal Image Mounter should support all image formats supported by DiscUtils, because it uses DiscUtils to parse the image files. Did you get any error messages when you tried? Or what more exactly happened?

**Comment by Keops10** [comment]
https://ibb.co/xCMfG78    https://ibb.co/C63d6rk   Yes, it shows as added, even in disk manager, but I can't see it on my computer.

**Comment by LTRData** [comment]
> https://ibb.co/xCMfG78 https://ibb.co/C63d6rk Yes, it shows as added, even in disk manager, but I can't see it on my computer.

Is it really a partition scheme and file system that is supported by Windows? I mean, if you would have attached a physical disk with the same contents to a Windows machine, would it have been possible to browse the file systems in Windows in that case?

If it is not a Windows compatible file system in your images, you cannot directly mount the images as virtual disks in this way in Windows. You would need some kind of file system emulation layer as well. I have a project like that over at my fork of dokan-dotnet. It uses DiscUtils libraries not only to parse disk contents but also to access the file systems in the disk images and then it uses Dokan virtual file system driver to mount them as virtual file systems. They will not show up as disks in Disk Management and not accessible to applications looking for physical disks in any way, but it is possible to browse such file systems in Explorer in Windows.
https://github.com/LTRData/dokan-dotnet/tree/LTRData.DokanNet-initial/DiscUtilsFs

**Comment by Keops10** [comment]
Thank you very much for writing so descriptively. Also, thank you for your interest. I think I understood where the problem is. As you said yes, if windows is a drive, it can display it as a disk. Since my vhdx is in Ext4 format, this is how I see it. I did some research in it and found the SharpExt4 library, but I'm not sure it will work. Sorry for my bad english.

**Comment by Ladwaldez** [comment]
Can you show me how you can do it

**Comment by Keops10** [comment]
While Vhdx can be read with DiscUtils, shouldn't I export the contents?

**Comment by LTRData** [comment]
> Can you show me how you can do it

If you build DiscUtilsFs from that dokan-dotnet fork, you can do:
`DiscUtilsFs -where=D: -vhd=c:\path\to\image.vdi`

You need to install Dokan first:
https://github.com/dokan-dev/dokany/releases

**Comment by Ladwaldez** [comment]
I downloaded the doc but I couldn't include it in my project. Seems too complicated to me. Here is my code where I can read vhdx with DiscUtils. With dokan, how can I make a change in this code, I can both read and export the directory I selected from the listbox. I really need this.

```
private void Mountet_Click(object sender, RoutedEventArgs e)
{
//Bir Linux ext4 disk görüntüsü aç
var disk = ExtDisk.Open(1);
//Linux bölümünü alın ve açın
var fs = ExtFileSystem.Open(disk, disk.Partitions[0]);
///etc klasöründeki tüm dosyaları listele
foreach (fs.GetFiles("/etc", "*", SearchOption.AllDirectories içindeki var dosyası))
{
Console.WriteLine(file);
}
} }
```

**Comment by LTRData** [comment]
Dokan is actually quite simple. The DiscUtilsFs project I linked to shows how to do it. You basically wrap a DiscUtils `IFileSystem` instance in a `DokanDiscUtils` object and then call Dokan API to mount it as a virtual file system. But you need to install Dokan driver too, that is why I linked to Dokan release pages too. You do not need to build it, just install it.

On the other hand, if all that you are looking for here is to copy files from a file system recognized by DiscUtils libraries, you do not really need a virtual file system or a drive letter visible in Windows Explorer. You could simply open each file in your `foreach` loop there like `fs.OpenFile(name, FileMode.Open, FileAccess.Read)` and the create a target file somewhere else and call `CopyTo` to copy source file data to destination file.

**Comment by Ladwaldez** [comment]
I added the DokanNet reference to my project. and I tried to do as follows, do you think it's right?


```
 var disk = new DiscUtils.Vhdx.Disk(secilen.Text, System.IO.FileAccess.Read);

                    DiscUtils.FileSystems.SetupHelper.SetupFileSystems();
                    var manager = new VolumeManager(disk);
                    var logicalVolumes = manager.GetLogicalVolumes();
                    foreach (var volume in logicalVolumes)
                    {
                        var fsInfos = FileSystemManager.DetectFileSystems(volume);

                        foreach (var fsInfo in fsInfos)
                        {
                            var fs = fsInfo.Open(volume);
                            
                            string[] array2 = fs.GetFiles(fs.Root.Name, "*.*", SearchOption.AllDirectories);
                            foreach (string name in array2)
                            {
                                secilenmakineanaliz.Items.Add(name);
                                fs.OpenFile(name, FileMode.Open, FileAccess.Read);
                                System.IO.File.Copy(name, @"D:\", true);
                            }
                        }
                    }
```

**Comment by LTRData** [comment]
Change
```
fs.OpenFile(name, FileMode.Open, FileAccess.Read);
System.IO.File.Copy(name, @"D:\", true);
```

To
```
using var sourceStream = fs.OpenFile(name, FileMode.Open, FileAccess.Read);
using var targetStream = System.IO.File.OpenWrite(@"D:\" + name);
sourceStream.CopyTo(targetStream);
```

You don't need Dokan for this. It does not mount as virtual disk or file system in any way. It just reads and copies data directly.

**Comment by Keops10** [comment]
dude you are amazing thank you for helping me patiently. Finally, I have one more question to ask var disk = new DiscUtils.Xva.Disk(secilen.Text, System.IO.FileAccess.Read);

When I make DiscUtils.Xva here, I see that the code is not working, where do you think I am going wrong? Also I need to do this in iso and img.

**Comment by LTRData** [comment]
> dude you are amazing thank you for helping me patiently. Finally, I have one more question to ask var disk = new DiscUtils.Xva.Disk(secilen.Text, System.IO.FileAccess.Read);
> 
> When I make DiscUtils.Xva here, I see that the code is not working, where do you think I am going wrong? Also I need to do this in iso and img.

You should really avoid specifying class of virtual disk in this way in your use case.

Use the methods described here to use all available format support:
https://github.com/DiscUtils/DiscUtils

Then, `VirtualDisk.OpenDisk` etc methods that will find correct library automatically.

**Comment by Keops10** [comment]
I want to display the structure I read in the treeview content. But I found that it doesn't read folder by folder in the loop. How can I do that.I can't thank you enough.

**Comment by LTRData** [comment]
> I want to display the structure I read in the treeview content. But I found that it doesn't read folder by folder in the loop. How can I do that.I can't thank you enough.

There are other methods like `GetDirectories` instead of `GetFiles` for that. You probably need to go down each directory and enumerate files there too etc.

**Comment by Keops10** [comment]
I am using Xaml. How can I number each directory, can you show me on my sample code above? It returns a directory instead of a folder in the loop. I guess my logic wasn't enough :(

**Comment by LTRData** [comment]
> I am using Xaml. How can I number each directory, can you show me on my sample code above? It returns a directory instead of a folder in the loop. I guess my logic wasn't enough :(

Sorry, I cannot write the entire program for you. I think you need to ask someone more familiar with XAML/WPF etc for such questions. It is a bit out-of-scope for DiscUtils.

**Comment by Keops10** [comment]
> 

Hello, with the method you have shown, I can easily copy a virtual machine image in windows structure to windows. The problem is that when I want to copy the contents of an image with a linux image to windows, I encounter errors due to filenames (* ? | < > :) how can I solve this?

**Comment by EricZimmerman** [comment]
Search for those illegal Characters and replace them on the new name

**Comment by Keops10** [comment]
yes i tried that. 'replace'
But the line I'm having trouble with is :fs.OpenFile(name, FileMode.Open, FileAccess.Read);

I guess it can't read because I read it on windows

**Comment by EricZimmerman** [comment]
Ext? Install a driver for ext and then mount it normally and things should just work to copy from what I recall

**Comment by Keops10** [comment]
You're right, but to speed things up, I want to get around this problem without mounting.

**Comment by Keops10** [comment]
I even took the risk of incomplete copying. fs.OpenFile(name.Replace('*', '_'), FileMode.Open, FileAccess.Read); I tried 'catch (FileNotFoundException)' but the result did not change. It seems to be working without any errors but I've checked that it's not copying. the problem persists.

**Comment by Ladwaldez** [comment]
I solved the problem with Path.GetPathRoot

**Comment by Keops10** [comment]
Can I copy the image without decompressing?

**Comment by LTRData** [comment]
We have made lots of changes in the LTRData fork of DiscUtils regarding file name and symbolic link compatibility. Have you tried it and it does not work there either for some reason, I can take a look at why and how it happens and make a fix for it.

**Comment by Keops10** [comment]
Thanks for the fix, I really need it. How soon will you fix it? Also, if you have the opportunity to try, can you try to read an image with .vmdk extension, is there a problem with my vmdk?

**Comment by LTRData** [comment]
No, I meant that you could try first with the LTRData.DiscUtils fork of this repository. That will probably help because I recognize this kind of issues as something we had when we developed File system bypass mount mode in Arsenal Image Mounter. We managed to fix most of it by emulating slightly changed file names etc to make it work on Windows systems. But then if you see similar issues when using that fork too, then I can try to fix it.

**Comment by Ladwaldez** [comment]
`DiscUtils.FileSystems.SetupHelper.SetupFileSystems();`
What should I use instead in LTRData

**Comment by Keops10** [comment]
sourceStream.CopyTo(targetStream);   > System.IO.IOException: 'Unable to find extent for block 0'
I'm getting an error and the problem persists when I pass this error with try I have the same problem in LTRData (I installed it as nuget)

**Comment by Keops10** [comment]
Yes I've seen LTRData fix special characters. Strangely, I found it stuck on a file with the .version extension, and when I told it to skip it, it worked fine.

**Comment by LTRData** [comment]
> sourceStream.CopyTo(targetStream);   > System.IO.IOException: 'Unable to find extent for block 0'
> I'm getting an error and the problem persists when I pass this error with try I have the same problem in LTRData (I installed it as nuget)

That looks interesting. Do you know what kind of file that was? Anything special with it? Is it an image that you could share somewhere?

**Comment by Keops10** [comment]
https://ibb.co/wsTW9fM  Of course... I guess it has something to do with null content.

**Comment by Keops10** [comment]
Then DiscUtils.Vmdk.Disk(textbox.Text, System.IO.FileAccess.Read); I detected that it is not reading the vmdk image.

**Comment by LTRData** [comment]
> https://ibb.co/wsTW9fM Of course... I guess it has something to do with null content.

Sorry for the confusion! I meant, the image file that you are trying to read files from, is that an image that you could share somewhere?

**Comment by Keops10** [comment]
If I can reach you privately, this could be it.

**Comment by LTRData** [comment]
> If I can reach you privately, this could be it.

Yes no problem. E-mail me at olof@ltr-data.se

**Comment by Keops10** [comment]
okey. I sent an e-mail.
